proposals/003-authentication-api.md

```diff
@@ -0,0 +1,522 @@
+This would be driven by the runtime following a successful TCP handshake on the client-to-proxy connection, and after `ClientSubjectAware` plugins have initially been called with any `X500Principal`, but before any SASL exchange on the client side.
+2. That the SASL initiator plugin should perform SASL authentication _lazily_, as soon as a KRPC requests propagates along the chain to the initator plugin.
+In this case the `ClientSubjectAware` plugins may have been called with a SASL principal.
+
+Supporting the former case would allow the broker's principal (as obtained by a `ServerSubjectAware` implementing plugin) to be used in the client-facing SASL exchange. This could be relevant for Kafka clients which validated or made use of the authenticate server (in this case proxy) principal. However, we're not aware of any clients that actually do this, so we won't consider it further.
+
+Supporting the latter case allows the proxy's server-facing SASL credentials to depend on the client-facing principal.
+Because of variation in behaviour of client libraries there is not a single type of request which will always be the first. 
+However,the existing `Sasl(Handshake|Authenticate)(Request|Response)Filter` interfaces can be used for the KRPC level implementation.
+Such a filter implementation needs to keep some kind of `seenFirstRequest` state if it wants to use `FilterContent.sendRequest()` to insert its own initial requests prior to forwarding requests from the client.
+
+What's missing is a way for the initiator plugin to inform other plugins of the outcome. For this purpose we will add the following two methods to the existing `FilterContext` interface in the `io.kroxylicious.proxy.filter` package.
+
+```java
+package io.kroxylicious.proxy.filter;
+
+public interface FilterContext {
+
+    // ... existing methods ...
+
+   
+    /**
+     * Allows a filter (typically one which implements {@link SaslAuthenticateRequestFilter})
+     * to announce a successful authentication outcome to subsequent plugins.
+     * @param subject The authenticated subject.
+     */
+    void serverAuthenticationSuccess(Subject subject);
+
+    /**
+     * Allows a filter (typically one which implements {@link SaslAuthenticateRequestFilter})
+     * to announce a failed authentication outcome to subsequent plugins.
+     * @param exception An exception describing the authentication failure. 
+     */
+    void serverAuthenticationFailure(LoginException exception);
+    
+}
+```
+
+These are the mirror image of the equivalent client methods, and calling them would similarly result in `ServerSubjectAware`-implememting plugins getting informed of the server's subject.
+
+
+### Use cases
+
+* SASL-to-SASL identity mapping (in combination with a SASL terminator)
+* SASL-to-SASL identity preservation with in-proxy authorization (in combination with a SASL terminator, and an authorizer).
+* Audit logging
```

Comment:
Adding another one which is my motivation for joining this discussion:

```suggestion
* Audit logging
* Projections in fetch responses based on client identity, i.e. only returning those fields of the records on a given topic a client is authorized for
```

Reply:
I'd call this "principal aware data masking", and yes, absolutely this is a use case.